uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,120 @@
+        self.enable_beta = enable_beta
+
+
+class FullAutoAttachResult(DataObject, AdditionalInfo):
+    pass
+
+
+def get_auto_attach_token(cfg: UAConfig, cloud: AutoAttachCloudInstance):
+    contract_client = contract.UAContractClient(cfg)
+    try:
+        tokenResponse = contract_client.request_auto_attach_contract_token(
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def is_any_beta(cfg: UAConfig, ents: List[str]) -> bool:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False
+
+
+def full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    instance = None  # type: Optional[AutoAttachCloudInstance]
+    token = None  # Optional[str]
+    instance = actions.get_cloud_instance(cfg)
+    allow_enable = not any([options.enable, options.enable_beta])
+    token = get_auto_attach_token(cfg, instance)  # type: ignore
+    actions.attach_with_token(cfg, token=token, allow_enable=allow_enable)
+    if allow_enable:
+        return FullAutoAttachResult()
+
+    services = set()
+    if options.enable:
+        if is_any_beta(cfg, options.enable):
+            raise exceptions.BetaServiceError(
+                msg="beta service found in the enable list",
+                msg_code="beta-service-found",
+            )
+        services.update(options.enable)
+    if options.enable_beta:
+        services.update(options.enable_beta)
+
+    services_ = list(set(services))
+    found, not_found = get_valid_entitlement_names(services_, cfg)
+    enabled_services = 0
+
+    for name in found:
+        ent_ret, reason = actions.enable_entitlement_by_name(
+            cfg, name, assume_yes=True, allow_beta=True
+        )
+        if not ent_ret:
+            if (
+                reason is not None
+                and isinstance(reason, CanEnableFailure)
+                and reason.message is not None
+            ):
+                raise exceptions.EntitlementNotEnabledError(
+                    msg=reason.message.msg,
+                    msg_code=reason.message.name,
+                    additional_info={"service": name},
+                )
+        else:
+            enabled_services += 1
+
+    if not_found:
+        msg = cli._create_enable_entitlements_not_found_message(
+            not_found, cfg=cfg, allow_beta=True
+        )
+        raise exceptions.EntitlementNotFoundError(msg.msg)
+    if len(services_) == enabled_services:
+        return FullAutoAttachResult()
+
+    raise exceptions.FullAutoAttachFailureError()
```

Comment:
We have the data to raise a better error here. Let's include which services failed to enable in the error as additional_info

Reply:
We raise an error when to fail to enable an entitlement i.e. `EntitlementNotEnableError`. Infact, I do not think we will ever reach the `FullAutoAttachFailureError` since we would have failed before incase of an error.